libs/common/src/vault/services/totp.service.ts

```diff
@@ -1,27 +1,43 @@
-import { firstValueFrom, map } from "rxjs";
+import { Observable, map, shareReplay, switchMap, timer } from "rxjs";
 
 import { TotpResponse } from "@bitwarden/sdk-internal";
 
 import { SdkService } from "../../platform/abstractions/sdk/sdk.service";
 import { TotpService as TotpServiceAbstraction } from "../abstractions/totp.service";
 
+/**
+ * Represents TOTP information including display formatting and timing
+ */
+export type TotpInfo = {
+  /** The TOTP code value */
+  totpCode: string;
+
+  /** The TOTP code value formatted for display, includes spaces */
+  totpCodeFormatted: string;
+
+  /** Progress bar percentage value */
+  totpDash: number;
+
+  /** Seconds remaining until the TOTP code changes */
+  totpSec: number;
+
+  /** Indicates when the code is close to expiring */
+  totpLow: boolean;
+};
+
 export class TotpService implements TotpServiceAbstraction {
   constructor(private sdkService: SdkService) {}
 
-  async getCode(key: string | undefined): Promise<TotpResponse | undefined> {
-    if (!key) {
-      return undefined;
-    }
-
-    return await firstValueFrom(
-      this.sdkService.client$.pipe(
-        map((sdk) => {
-          if (sdk === undefined) {
-            throw new Error("SDK is undefined");
-          }
-          return sdk.vault().totp().generate_totp(key);
-        }),
+  getCode$(key: string): Observable<TotpResponse> {
+    return timer(0, 1000).pipe(
+      switchMap(() =>
+        this.sdkService.client$.pipe(
+          map((sdk) => {
+            return sdk.vault().totp().generate_totp(key);
+          }),
+        ),
       ),
+      shareReplay({ refCount: true, bufferSize: 1 }),
```

Comment:
suggestion(non-blocking): you could also filter out emissions that return the same code using `distinctUntilChanged`. You'd probably have to add a manual comparator because the SDK would not be returning the same object

```suggestion
      distinctUntilChanged((prev, curr) => prev.code === curr.code)
      shareReplay({ refCount: true, bufferSize: 1 }),
```